spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java

```diff
@@ -0,0 +1,399 @@
+			this.leaderEventPublisher = new DefaultLeaderEventPublisher(
+					this.applicationEventPublisher);
+		}
+		synchronized (this.lifecycleMonitor) {
+			if (!this.running) {
+				this.leaderSelector = new LeaderSelector(buildLeaderPath());
+				this.future = this.executorService.submit(this.leaderSelector);
+				this.running = true;
+				logger.debug("Started LeaderInitiator");
+			}
+		}
+	}
+
+	/**
+	 * Stop the registration of the {@link #candidate} for leader election. If the
+	 * candidate is currently leader, its leadership will be revoked.
+	 */
+	@Override
+	public void stop() {
+		synchronized (this.lifecycleMonitor) {
+			if (this.running) {
+				this.running = false;
+				this.future.cancel(true);
+				logger.debug("Stopped LeaderInitiator");
+			}
+		}
+	}
+
+	@Override
+	public void stop(Runnable runnable) {
+		stop();
+		runnable.run();
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		stop();
+		this.executorService.shutdown();
+	}
+
+	/**
+	 * Sets the {@link LeaderEventPublisher}.
+	 *
+	 * @param leaderEventPublisher the event publisher
+	 */
+	public void setLeaderEventPublisher(LeaderEventPublisher leaderEventPublisher) {
+		this.leaderEventPublisher = leaderEventPublisher;
+	}
+
+	/**
+	 * @return the lock key used by leader election
+	 */
+	private String buildLeaderPath() {
+		return this.candidate.getRole();
+	}
+
+	/**
+	 * @return the context (or null if not running)
+	 */
+	public Context getContext() {
+		if (this.leaderSelector == null) {
+			return NULL_CONTEXT;
+		}
+		return this.leaderSelector.context;
+	}
+
+	class LeaderSelector implements Callable<Void> {
+
+		private final Lock lock;
+		private final LockContext context = new LockContext();
+		private volatile boolean locked = false;
+
+		LeaderSelector(String lockKey) {
+			this.lock = LockRegistryLeaderInitiator.this.locks.obtain(lockKey);
+		}
+
+		@Override
+		public Void call() throws Exception {
+			try {
+				while (LockRegistryLeaderInitiator.this.running) {
+					try {
+						// We always try to acquire the lock, in case it expired
+						boolean acquired = this.lock.tryLock(LockRegistryLeaderInitiator.this.heartBeatMillis,
+								TimeUnit.MILLISECONDS);
+						if (!this.locked) {
+							if (acquired) {
+								// Success: we are now leader
+								this.locked = true;
+								LockRegistryLeaderInitiator.this.candidate.onGranted(this.context);
+								if (LockRegistryLeaderInitiator.this.leaderEventPublisher != null) {
+									LockRegistryLeaderInitiator.this.leaderEventPublisher.publishOnGranted(
+											LockRegistryLeaderInitiator.this, this.context,
+											LockRegistryLeaderInitiator.this.candidate.getRole());
+								}
+							}
+						}
+						else if (acquired) {
+							// If we were able to acquire it but we were already locked we
+							// should release it
+							this.lock.unlock();
+							// Give it a chance to expire.
+							Thread.sleep(LockRegistryLeaderInitiator.this.heartBeatMillis);
```

Comment:
Doh. I missed the fact that we re-acquire the lock and have `holdCount` there and don't unlock in the target store:
```
if (this.delegate.getHoldCount() > 1) {
	this.delegate.unlock();
	return;
}
try {
	this.mutex.delete(this.path);
}
```
And right: after using different `LockRepository` my test starts to fail as expected.

Now the algorithm becomes more robust for me :smile: